cranelift-codegen/src/regalloc/reload.rs

```diff
@@ -211,6 +211,46 @@ impl<'a> Context<'a> {
         debug_assert!(self.candidates.is_empty());
         self.find_candidates(inst, constraints);
 
+        // If we find a copy from a stack slot to the same stack slot, replace
+        // it with a `copy_nop` but otherwise ignore it.  In particular, don't
+        // generate a reload immediately followed by a spill.  The `copy_nop`
+        // has a zero-length encoding, so will disappear at emission time.
+        if let InstructionData::Unary {
+            opcode: Opcode::Copy,
+            arg,
+        } = self.cur.func.dfg[inst]
+        {
+            let dst_vals = self.cur.func.dfg.inst_results(inst);
+            if dst_vals.len() == 1 {
+                let dst_val = dst_vals[0];
+                let src_loc = self.cur.func.locations[arg];
+                let dst_loc = self.cur.func.locations[dst_val];
+                let src_aff = self.liveness[arg].affinity;
+                let dst_aff = self.liveness[dst_val].affinity;
+                let can_transform = match (src_loc, dst_loc) {
+                    (ValueLoc::Stack(src_slot), ValueLoc::Stack(dst_slot)) => {
+                        src_slot == dst_slot && src_aff.is_stack() && dst_aff.is_stack()
```

Comment:
Can you explain why we need to at the values' affinities? Even if a value is supposed to live in a register and is not (because of a diversion, I imagine?), I can't imagine a reason why we couldn't avoid the spill.

Reply:
Hmm, excess paranoia on my part.  I removed it.  On reading the comments for `Affinity` I see it's not relevant here.